src/dispatch/case/flows.py

```diff
@@ -608,107 +610,178 @@ def case_assign_role_flow(
-            storage_members = [case.tactical_group.email]
-        # direct add members if not group exists
-        else:
-            storage_members = direct_participant_emails
+    @check_resource_exists("storage")
+    def create_storage(self, individual_participants: ParticipantEmails) -> None:
+        """Create a storage resource for the case if it does not exist."""
+        storage_members = [p.email for p in individual_participants]
+        if self.case.tactical_group:
+            storage_members.append(self.case.tactical_group.email)
 
-        if not case.storage:
-            storage_flows.create_storage(
-                subject=case, storage_members=storage_members, db_session=db_session
-            )
+        storage_flows.create_storage(
+            subject=self.case, storage_members=storage_members, db_session=self.db_session
+        )
 
-        # we create the investigation document
-        if not case.case_document:
-            document_flows.create_document(
-                subject=case,
-                document_type=DocumentResourceTypes.case,
-                document_template=case.case_type.case_template_document,
-                db_session=db_session,
-            )
+    @check_resource_exists("document")
+    def create_document(self) -> None:
+        """Create a document for the case if it does not exist."""
+        if self.case.case_document:
+            raise AlreadyExists(resource_name="Document", message="Case document already exists.")
+
+        document_flows.create_document(
+            subject=self.case,
+            document_type=DocumentResourceTypes.case,
+            document_template=self.case.case_type.case_template_document,
+            db_session=self.db_session,
+        )
 
-        # we update the case document
+    def update_document(self) -> None:
+        """Update the case document."""
         document_flows.update_document(
-            document=case.case_document, project_id=case.project.id, db_session=db_session
+            document=self.case.case_document,
+            project_id=self.case.project.id,
+            db_session=self.db_session,
         )
 
-    try:
-        # we create the conversation and add participants to the thread
-        conversation_flows.create_case_conversation(case, conversation_target, db_session)
-
-        event_service.log_case_event(
-            db_session=db_session,
-            source="Dispatch Core App",
-            description="Conversation added to case",
-            case_id=case.id,
-        )
-        # wait until all resources are created before adding suggested participants
-        individual_participants = [x.email for x, _ in individual_participants]
+    @check_resource_exists("conversation")
+    def create_conversation(
+        self, conversation_target: str | None, individual_participants: ParticipantEmails
+    ) -> None:
+        """Create a conversation thread for the case and add participants."""
 
```

Comment:
Should we `raise AlreadyExists` if it exists?

Reply:
The decorator raises that error, if you're referring to the resource:

https://github.com/Netflix/dispatch/pull/4164/files#diff-ee560a5825ccb96768d929d3eb4a3cc23526ce20e31155e79dda5180e9065f0bR618